dhall/tests/Dhall/Test/ETags.hs

```diff
@@ -0,0 +1,70 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Dhall.Test.ETags where
+
+import Data.Monoid ((<>))
+import Data.Text (Text)
+import Dhall.Util (Input(..))
+import Prelude hiding (FilePath)
+import Test.Tasty (TestTree)
+import Turtle (FilePath)
+
+import qualified Data.Text        as Text
+import qualified Data.Text.IO     as Text.IO
+import qualified Dhall.ETags      as ETags
+import qualified Dhall.Test.Util  as Test.Util
+import qualified Test.Tasty       as Tasty
+import qualified Test.Tasty.HUnit as Tasty.HUnit
+import qualified Turtle
+
+etagsDirectory :: FilePath
+etagsDirectory = "./tests/etags"
+
+getTests :: IO TestTree
+getTests = do
+    etagsTests <- Test.Util.discover (Turtle.chars <* ".dhall") etagsTest (Turtle.lstree etagsDirectory)
+
+    let testTree = Tasty.testGroup "etags tests" [ etagsTests, etagsDirTest ]
+
+    return testTree
+
+etagsTest :: Text -> TestTree
+etagsTest prefix =
+    Tasty.HUnit.testCase (Text.unpack prefix) $ do
+        let inputFile  = Text.unpack (prefix <> ".dhall")
+        let outputFile = Text.unpack (prefix <> ".tags")
+
+        actualTags <- fixPathSeparators <$> ETags.generate (InputFile inputFile) Nothing False
+
+        expectedTags <- readExpected outputFile
+
+        let message = "The actual tags did not match the expected tags"
+
+        Tasty.HUnit.assertEqual message expectedTags actualTags
+
+etagsDirTest :: TestTree
+etagsDirTest =
+    Tasty.HUnit.testCase "all" $ do
+        let outputFile = Text.unpack . Turtle.format Turtle.fp $ etagsDirectory Turtle.</> "all.tags"
+
+        actualTags <- fmap fixPathSeparators
+                      (ETags.generate
+                          (InputFile (Text.unpack . Turtle.format Turtle.fp $ etagsDirectory))
+                          (Just [".dhall"])
+                          False)
+
+        expectedTags <- readExpected outputFile
+
+        let message = "The actual tags did not match the expected tags for directory test"
+
+        Tasty.HUnit.assertEqual message expectedTags actualTags
+
+readExpected :: String -> IO Text
+readExpected file = fixPathSeparators <$> Text.IO.readFile file
+
+fixPathSeparators :: Text -> Text
+fixPathSeparators t = Text.intercalate "\n" (map fix prevLineLine)
+    where ls = Text.lines t
+          prevLineLine = zip ("":ls) ls
+          fix ("\x0c", l) = Text.replace "\\" "/" l
+          fix (     _, l) = l
```

Comment:
Test code should be simple IMHO. This code isn't and it doesn't say what it does.

Please return to the old definition, and apply the function only to the output of `generate`.

Let's just leave it to the people who add the next tests to use `/` for path separators.

Reply:
I agree about test code. It's better to have as simple as it's possible. Just wanted to prevent confusing situation in the future. On the other hand, it can make development more conscious, because it's less magic.
Done.